futures-executor/src/current_thread.rs

```diff
@@ -0,0 +1,158 @@
+use futures_core::task::{self, NotifyHandle};
+use futures_core::{Future, Async};
+
+use thread::ThreadNotify;
+use task_runner::{set_current, with_current, TaskRunner};
+
+/// Provides execution context
+///
+/// This currently does not do anything, but allows future improvements to be
+/// made in a backwards compatible way.
+pub struct Context<'a> {
+    // enter: &'a executor::Enter,
+    runner: &'a mut TaskRunner,
+    handle: &'a (Fn() -> NotifyHandle + 'a),
+    thread: &'a ThreadNotify,
+}
+
+pub fn run<F, R>(f: F) -> R
+    where F: FnOnce(&mut Context) -> R
+{
+    ThreadNotify::with_current(|thread| {
+        let mut runner = TaskRunner::new();
+        let handle = &|| thread.clone().into();
+
+        // Kick off any initial work through the callback provided
+        let ret = set_current(&runner.executor(), || {
+            f(&mut Context {
+                // enter: &,
+                runner: &mut runner,
+                handle,
+                thread,
+            })
+        });
+
+        // So long as there's pending work we keep polling and sleeping.
+        if !runner.is_done() {
+            loop {
+                runner.poll(handle);
+                if runner.is_done() {
+                    break
+                }
+                thread.park();
+            }
+        }
+
+        return ret
+    })
+}
+
+/// Spawns a future on the current thread.
+///
+/// The provided future must complete or be canceled before
+/// `run` will return.
+///
+/// # Panics
+///
+/// This function can only be invoked from the context of a
+/// `run` call; any other use will result in a panic.
+pub fn spawn<F>(future: F)
+    where F: Future<Item = (), Error = ()> + 'static
+{
+    _spawn(future, false).unwrap_or_else(|_| {
+        panic!("cannot call `execute` unless the thread is already \
+                in the context of a call to `run`")
+    })
+}
+
+/// Spawns a daemon future on the current thread.
+///
+/// Completion of the provided future is not required for the pending
+/// `run` call to complete. If `run` returns before `future` completes, it
+/// will be dropped.
+///
+/// # Panics
+///
+/// This function can only be invoked from the context of a
+/// `run` call; any other use will result in a panic.
+pub fn spawn_daemon<F>(future: F)
+    where F: Future<Item = (), Error = ()> + 'static
+{
+    _spawn(future, true).unwrap_or_else(|_| {
+        panic!("cannot call `execute` unless the thread is already \
+                in the context of a call to `run`")
+    })
+}
+
+/// Cancels *all* executing futures.
+///
+/// This cancels both daemon and non-daemon futures.
+///
+/// # Panics
+///
+/// This function can only be invoked from the context of a
+/// `run` call; any other use will result in a panic.
+pub fn cancel_all_executing() {
+    with_current(|current| {
+        match current {
```

Comment:
Again WRT `expect` question-- is there a reason why this isn't `current.expect("...").cancel_all_executing();`?